ls1intum/Artemis
# `General`: Personal Access Tokens (PAT)

### Checklist
#### General


- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server. (KIT Test System)
#### Server
#### Client


### Motivation and Context


Since we are developing applications for Artemis like [eclipse-artemis](https://github.com/kit-sdq/programming-lecture-eclipse-artemis), we would like to have an authentication workflow that is independent from the used login mechanism.
E.g., we at KIT are currently using the SAML2/Shibboleth workflow for registering users & the internal user management of Artemis.

Since our Eclipse Plugin directly talks to the REST API of Artemis, we currently rely on the ability to login via username and password to get an JWT Token. Nevertheless, we want to disable the ability to change the password of users in Artemis and want to enable https://github.com/ls1intum/Artemis/pull/4749 to clone repositories. This would make our Grading Plugin not usable, since it could not communicate with the API anymore.

@dfuchss 

Implementation using JWT Tokens for #5075.
Closes #5075

### Description
This PR adds a new REST endpoint that allows a logged-in user to generate new JWT Tokens that can be used as personal access tokens. The feature can be disabled using feature toggles.

### Steps for Testing

Prerequisites:
None

0. Activate "pat" profile. You can use prebuild docker images from https://github.com/kit-sdq/Artemis4Docker/pkgs/container/artemis/27516500?tag=pr-5216
1. Log in to Artemis
2. Navigate to Settings ⇾ Personal Access Tokens
3. Press Generate
4. Check that the JWT Token works for the API (e.g., using Postman)

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|


### Screenshots


![image](https://user-images.githubusercontent.com/8215084/173060309-fe738552-ef86-4c74-bb6f-cf3452a7309f.png)
![image](https://user-images.githubusercontent.com/8215084/173060347-e74a5d24-6f39-422b-9db7-0e57f27eb22e.png)
